src/main/java/io/numaproj/pulsar/consumer/PulsarSource.java

```diff
@@ -7,139 +7,111 @@
-        if (!messages.isEmpty()) {
+        // If there are messages not acknowledged, return
+        if (!messagesToAck.isEmpty()) {
+            log.trace("messagesToAck not empty: {}", messagesToAck);
             return;
         }
 
-        for (int i = 0; i < request.getCount(); i++) {
-            if (System.currentTimeMillis() - startTime > request.getTimeout().toMillis()) {
-                return;
-            }
+        Consumer<byte[]> consumer = null;
 
-            try {
-                org.apache.pulsar.client.api.Message<byte[]> pMsg = pulsarConsumer
-                        .receive((int) request.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
-                if (pMsg == null) {
-                    return;
-                }
+        try {
+            // Obtain a consumer with the desired settings.
+            consumer = pulsarConsumerManager.getOrCreateConsumer(request.getCount(), request.getTimeout().toMillis());
 
-                log.info("Consumed Pulsar message: {}", new String(pMsg.getValue()));
+            Messages<byte[]> batchMessages = consumer.batchReceive();
 
-                byte[] offsetBytes = pMsg.getMessageId().toString().getBytes(StandardCharsets.UTF_8);
-                Offset offset = new Offset(offsetBytes);
+            if (batchMessages == null || batchMessages.size() == 0) {
+                log.trace("Received 0 messages, return early.");
+                return;
+            }
 
-                Map<String, String> headers = new HashMap<>();
-                headers.put("pulsarMessageId", pMsg.getMessageId().toString());
+            // Process each message in the batch.
+            for (org.apache.pulsar.client.api.Message<byte[]> pMsg : batchMessages) {
+                String msgId = pMsg.getMessageId().toString();
+                log.info("Consumed Pulsar message [id: {}]: {}", pMsg.getMessageId(),
+                        new String(pMsg.getValue(), StandardCharsets.UTF_8));
 
-                Message message = new Message(pMsg.getValue(), offset, Instant.now(), headers);
+                byte[] offsetBytes = msgId.getBytes(StandardCharsets.UTF_8);
+                Offset offset = new Offset(offsetBytes);
 
+                Message message = new Message(pMsg.getValue(), offset, Instant.now());
                 observer.send(message);
-                messages.put(pMsg.getMessageId().toString(), pMsg);
 
-            } catch (PulsarClientException e) {
-                log.error("Failed to receive message from Pulsar", e);
-                return;
+                messagesToAck.put(msgId, pMsg);
             }
+        } catch (PulsarClientException e) {
+            log.error("Failed to get consumer or receive messages from Pulsar", e);
+            throw new RuntimeException("Failed to get consumer or receive messages from Pulsar", e);
         }
     }
 
     @Override
     public void ack(AckRequest request) {
```

Comment:
please add a check here to compare the toAck map with AckRequest, if not 100% match, log.err and return early.